nevergrad/functions/gym/gym_multi.py

```diff
@@ -0,0 +1,663 @@
+                if type(a) == tuple:
+                    a = tuple(int(_a + 0.5) for _a in a)
+                else:
+                    for i in range(len(a)):
+                        a[i] = self.subaction_type(a[i])
+        assert type(a) == self.action_type, f"{a} should have type {self.action_type} "
+        try:
+            assert env.action_space.contains(
+                a
+            ), f"In {self.name}, high={env.action_space.high} low={env.action_space.low} {a} is not sufficiently close to {[env.action_space.sample() for _ in range(10)]}"
+        except AttributeError:
+            pass  # Not all env can do "contains".
+        return a
+
+    def step(self, a):
+        """Apply an action.
+
+        We have a step on top of Gym's step for storing some statistics."""
+        o, r, done, info = self.env.step(a)
+        return o, r, done, info
+
+    def heuristic(self, o, current_observations):
+        current_observations = np.asarray(current_observations + [o], dtype=np.float32)
+        self.archive = [
+            self.archive[i] for i in range(len(self.archive)) if self.archive[i][2] <= self.mean_loss
+        ]
+        self.archive = sorted(self.archive, key=lambda trace: -len(trace[0]))
+        for trace in self.archive:
+            to, ta, _ = trace
+            assert len(to) == len(ta)
+            if len(current_observations) > len(to) and "extrapolate" not in self.control:
+                continue
+            to = np.asarray(to[(-len(current_observations)) :], dtype=np.float32)
+            # if all((_to - _o) for _to, _o in zip(to, current_observations)) <= 1e-7:
+            if np.array_equal(to, current_observations):
+                return np.asarray(ta[len(current_observations) - 1], dtype=np.float32)
+        return None
+
+    def gym_simulate(self, x: np.ndarray, seed: int, pb_index: int, limited_fidelity: bool = True):
+        """Single simulation with parametrization x."""
+        current_time_index = 0
+        current_reward = 0.0
+        current_observations: tp.List[tp.Any] = []
+        current_actions: tp.List[tp.Any] = []
+        try:
+            if self.policy_shape is not None:
+                x = x.reshape(self.policy_shape)
+        except:
+            assert False, f"x has shape {x.shape} and needs {self.policy_shape} for control {self.control}"
+        assert seed == 0 or self.control != "conformant" or self.randomized
+        env = self.env
+        env.seed(seed=seed)
+        if "compilergym" in self.name:
+            if "stoc" in self.name:  # Stochastic case: in training we use randomly drawn points.
+                # In training, we randomly draw in csmith (but we are allowed to use 100x more budget :-) ).
+                o = env.reset(
+                    benchmark=np.random.choice(self.csmith) if pb_index < 0 else self.uris[pb_index]
+                )
+            else:
+                o = env.reset(benchmark=self.uris[self.compilergym_index])
```

Comment:
If I'm not wrong we want to simulate using a random problem in csmith for training and uris for testing. Is it happening there ? I don't think self.name changes so we are simulating using csmith all the time for stochasticCompilerGym. Am I missing something ?

Reply:
Maybe pb_index should be an env and not an int actually.

Reply:
Good idea it would make it much clearer